modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/AclRoutingSearchProcessor.java

```diff
@@ -0,0 +1,170 @@
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    @Override
+    public SearchRequest processRequest(SearchRequest request) throws Exception {
+        if (!extractFromQuery || request.source() == null) {
+            return request;
+        }
+
+        QueryBuilder query = request.source().query();
+        if (query == null) {
+            return request;
+        }
+
+        List<String> aclValues = extractAclValues(query);
+        if (aclValues.isEmpty()) {
+            return request;
+        }
+
+        // Generate routing values
+        String[] routingValues = aclValues.stream().map(this::generateRoutingValue).toArray(String[]::new);
+
+        // Set routing on the request
+        request.routing(routingValues);
+
+        return request;
+    }
+
+    private List<String> extractAclValues(QueryBuilder query) {
+        List<String> aclValues = new ArrayList<>();
+
+        if (query instanceof TermQueryBuilder) {
+            TermQueryBuilder termQuery = (TermQueryBuilder) query;
+            if (aclField.equals(termQuery.fieldName())) {
+                aclValues.add(termQuery.value().toString());
+            }
+        } else if (query instanceof TermsQueryBuilder) {
+            TermsQueryBuilder termsQuery = (TermsQueryBuilder) query;
+            if (aclField.equals(termsQuery.fieldName())) {
+                termsQuery.values().forEach(value -> aclValues.add(value.toString()));
+            }
+        } else if (query instanceof BoolQueryBuilder) {
+            BoolQueryBuilder boolQuery = (BoolQueryBuilder) query;
+
+            // Check must clauses
+            for (QueryBuilder mustClause : boolQuery.must()) {
+                aclValues.addAll(extractAclValues(mustClause));
+            }
+
+            // Check filter clauses
+            for (QueryBuilder filterClause : boolQuery.filter()) {
+                aclValues.addAll(extractAclValues(filterClause));
+            }
+
+            // Check should clauses if minimum_should_match > 0
+            if (boolQuery.minimumShouldMatch() != null && !boolQuery.should().isEmpty()) {
+                for (QueryBuilder shouldClause : boolQuery.should()) {
+                    aclValues.addAll(extractAclValues(shouldClause));
+                }
+            }
+        }
+
+        return aclValues;
+    }
+
+    private String generateRoutingValue(String aclValue) {
+        // Use MurmurHash3 for consistent hashing (same as ingest processor)
+        byte[] bytes = aclValue.getBytes(StandardCharsets.UTF_8);
+        MurmurHash3.Hash128 hash = MurmurHash3.hash128(bytes, 0, bytes.length, 0, new MurmurHash3.Hash128());
+
+        // Convert to base64 for routing value
+        byte[] hashBytes = new byte[16];
+        System.arraycopy(longToBytes(hash.h1), 0, hashBytes, 0, 8);
+        System.arraycopy(longToBytes(hash.h2), 0, hashBytes, 8, 8);
+
+        return Base64.getUrlEncoder().withoutPadding().encodeToString(hashBytes);
```

Comment:
Same comment as on the ingest side for the static variable.

Reply:
Fixed